File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int helper(int idx,vector<int>& nums,vector<int>&dp){
        if(idx==0){
            return nums[idx];
        }
        if(idx<0){
            return 0;
        }
        if(dp[idx]!=-1){
            return dp[idx];
        }
        int pick=nums[idx]+helper(idx-2,nums,dp);
        int notpic=0+helper(idx-1,nums,dp);
        return dp[idx]=max(pick,notpic);
    }
    int rob(vector<int>& nums) {
        int n=nums.size();
        vector<int>dp(n,-1);
        return helper(n-1,nums,dp);
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int helper(int[] nums,int n,int[] dp){
        if(n==0){
            return nums[n];
        }
        if(n<0){
            return 0;
        }

        if(dp[n]!=-1){
            return dp[n];
        }

        int pick=nums[n]+helper(nums,n-2,dp);
        int notpic=helper(nums,n-1,dp);
        return dp[n]=Math.max(pick,notpic);
    }
    public int rob(int[] nums) {
        int n=nums.length;
        int[] dp=new int[n];
        Arrays.fill(dp, -1);  
        return helper(nums,n-1,dp);
    }
}